standard/expressions.md

```diff
@@ -36,6 +36,142 @@ Most of the constructs that involve an expression ultimately require the express
+
+Certain combinations of static type of the pattern input value and the given type are considered incompatible and result in a compile-time error. A value of static type `E` is said to be ***pattern compatible*** with the type `T` if there exists an identity conversion, an implicit reference conversion, a boxing conversion, an explicit reference conversion, or an unboxing conversion from `E` to `T`, or if either `E` or `T` is an open type ([§9.4.3](types.md#943-open-and-closed-types)). It is a compile-time error if a value of type `E` is not pattern compatible with the type in a type pattern with which it is matched.
+
+> *Note*: The support for open types can be most useful when checking types that may be either struct or class types, and boxing is to be avoided. *end note*
+
+> *Example*: The declaration pattern is useful for performing run-time type tests of reference types, and replaces the idiom
+> ```csharp
+> var v = expr as Type;
+> if (v != null) { /* code using v */ }
+> ```
+> with the slightly more concise
+> ```csharp
+> if (expr is Type v) { /* code using v */ }
+> ```
+> *end example*
+
+It is an error if *type* is a nullable value type.
+
+> *Example*: The declaration pattern can be used to test values of nullable types: a value of type `Nullable<T>` (or a boxed `T`) matches a type pattern `T2 id` if the value is non-null and the type of `T2` is `T`, or some base type or interface of `T`. For example, in the code fragment
+> ```csharp
+> int? x = 3;
+> if (x is int v) { /* code using v */ }
+> ```
+> The condition of the `if` statement is `true` at runtime and the variable `v` holds the value `3` of type `int` inside the block. *end example*
+
+#### §constant-pattern-new-clause Constant pattern
+
+A *constant_pattern* is used to test the value of a pattern input value (§patterns-new-clause) against the given constant value.
+
+```antlr
+constant_pattern
+    : constant_expression
+    ;
+```
+
+Given a pattern input value *e* and a *constant_expression* *c*, if *e* and *c* have integral types, the pattern is considered matched if the result of the expression `e == c` is `true`. Otherwise, the pattern is considered matched if `object.Equals(e, c)` returns `true`. In this case it is a compile-time error if the static type of *e* is not pattern compatible (§declaration-pattern-new-clause) with the type of the constant.
```

Comment:
I think "In this case it is a compile-time error if the static type of *e* is not pattern compatible with the type of the constant" is backwards.  If you have a variable `e` of type `object` and test `e is "x"`, the static type of *e* (`object`) is not pattern compatible with the type of the constant (`string`).